services/121-service/test/performance/paginate-transactions-and-retry.test.ts

```diff
@@ -0,0 +1,146 @@
+    accessToken = await getAccessToken();
+    // Upload registration
+    const paymentId = await seedPaidRegistrations(
+      [registration],
+      programIdSafaricom,
+      10,
+      [TransactionStatusEnum.error],
+    );
+
+    // Change phone number
+    await updateRegistration(
+      programIdSafaricom,
+      registration.referenceId,
+      {
+        phoneNumber: '254708374149', // Phonenumber that does not fail
+      },
+      'test reason',
+      accessToken,
+    );
+    // Duplicate registrations
+    const duplicateRegistrationsResponse =
+      await mockRegistrationsAndPaymentData({
+        powerNumberRegistration: duplicateNumber,
+        numberOfPayments: 1,
+        accessToken,
+        body: {
+          secret: env.RESET_SECRET,
+        },
+      });
+    expect(duplicateRegistrationsResponse.statusCode).toBe(HttpStatus.CREATED);
+
+    // Get one page of transaction to test the duration of the api response
+    const getTransactionsStartTime = Date.now();
+    const paginatedTransactionsResponse = await getTransactions({
+      programId: programIdSafaricom,
+      accessToken,
+      page: 1,
+      limit: 10,
+      paymentId,
+      search: '3', // This is random filter to reduce result set, it seems likely that referenceIds contain '3'
+      filter: {
+        'filter.registrationReferenceId': '2', // This is random filter to reduce result set, it seems likely that referenceIds contain '2'
+      },
+    });
+    const getTransactionsElapsedTime = Date.now() - getTransactionsStartTime;
+
+    const twoSeconds = 2 * 1000;
+    expect(getTransactionsElapsedTime).toBeLessThan(twoSeconds);
+
+    expect(paginatedTransactionsResponse.statusCode).toBe(HttpStatus.OK);
+
+    // Get all transactions to simulate export
+    const supportedNumberOrRegistrations = 100_000; // Adjust based on expected supported number
+    const getAllTransactionsStartTime = Date.now();
+    const allTransactionsResponse = await getTransactions({
+      programId: programIdSafaricom,
+      accessToken,
+      paymentId,
+      limit: supportedNumberOrRegistrations,
+    });
+    const allTransactions = allTransactionsResponse.body.data;
+    expect(allTransactions.length).toBe(
+      Math.min(supportedNumberOrRegistrations, Math.pow(2, duplicateNumber)),
+    );
+    const getAllTransactionsElapsedTime =
+      Date.now() - getAllTransactionsStartTime;
+
+    const twoMinutes = 2 * 60 * 1000;
+    expect(getAllTransactionsElapsedTime).toBeLessThan(twoMinutes);
+
+    // Retry payment
+    const patchRetryRequestStartTime = Date.now();
+    const doPaymentResponse = await retryPayment({
+      programId: programIdSafaricom,
+      paymentId,
+      accessToken,
+      filter: { 'filter.status': `${TransactionStatusEnum.error}` },
+    });
+
+    const totalPayments = Math.pow(2, duplicateNumber);
+    const patchRetryRequestElapsedTime =
+      Date.now() - patchRetryRequestStartTime;
+    const fourMinutes = 4 * 60 * 1000;
```

Comment:
in our overall setup of these files we sort-of imply that DUPLICATE_NUMBER is a parameter that can be changed to anything. But with absolute values like this, that gets a bit iffy. Although I guess 17 is the upper limit, and thus 4 minutes is the upper limit. So also with lower values this will pass.. I guess it's fine for now.